File: platform/android/java/lib/src/com/crossbow/library/plugin/SignalInfo.java
package com.crossbow.library.plugin;

import android.text.TextUtils;
import java.util.Arrays;
import androidx.annotation.NonNull;
import com.crossbow.library.JNIUtil;

/**
 * Store information about a {@link CrossbowPlugin}'s signal.
 */
public final class SignalInfo {
	private final String name;
	private final Class<?>[] paramTypes;
	private final String[] paramTypesNames;

	public SignalInfo(@NonNull String signalName, Class<?>... paramTypes) {
		if (TextUtils.isEmpty(signalName)) {
			throw new IllegalArgumentException("Invalid signal name: " + signalName);
		}

		this.name = signalName;
		this.paramTypes = paramTypes == null ? new Class<?>[ 0 ] : paramTypes;
		this.paramTypesNames = new String[this.paramTypes.length];
		for (int i = 0; i < this.paramTypes.length; i++) {
			this.paramTypesNames[i] = JNIUtil.getJNIClassSignature(this.paramTypes[i]);
		}
	}

	public String getName() {
		return name;
	}

	Class<?>[] getParamTypes() {
		return paramTypes;
	}

	String[] getParamTypesNames() {
		return paramTypesNames;
	}

	@Override
	public String toString() {
		return "SignalInfo{"
				+
				"name='" + name + '\'' +
				", paramsTypes=" + Arrays.toString(paramTypes) +
				'}';
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) {
			return true;
		}
		if (!(o instanceof SignalInfo)) {
			return false;
		}

		SignalInfo that = (SignalInfo)o;

		return name.equals(that.name);
	}

	@Override
	public int hashCode() {
		return name.hashCode();
	}
}

File: platform/android/java/lib/src/com/crossbow/library/plugin/SignalInfo.kt
package com.crossbow.library.plugin

import java.lang.Class
import android.text.TextUtils
import java.lang.IllegalArgumentException
import com.crossbow.library.JNIUtil
import java.util.Arrays

/**
 * Store information about a [CrossbowPlugin]'s signal.
 */
class SignalInfo(signalName: String, vararg argParamTypes: Class<*>) {
    val name: String
    val paramTypes: Array<Class<*>>
    val paramTypesNames: Array<String>

    init {
        require(!TextUtils.isEmpty(signalName)) { "Invalid signal name: $signalName" }
        name = signalName
        paramTypes = arrayOf(*argParamTypes)
        val tmpArray = arrayOfNulls<String>(paramTypes.size)
        for (i in paramTypes.indices) {
            val tmp = JNIUtil.getJNIClassSignature(paramTypes[i])
            if (tmp !== null) {
                tmpArray[i] = tmp
            }
        }
        paramTypesNames = tmpArray.filterNotNull().toTypedArray()
    }

    override fun toString(): String {
        return ("SignalInfo{"
                +
                "name='" + name + '\'' +
                ", paramsTypes=" + Arrays.toString(paramTypes) +
                '}')
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o !is SignalInfo) {
            return false
        }
        return name == o.name
    }

    override fun hashCode(): Int {
        return name.hashCode()
    }
}
